[SPARK-16282][SQL] Implement percentile SQL function.

jiangxb1987 opened the pull request:
## What changes were proposed in this pull request?

Implement percentile SQL function. It computes the exact percentile(s) of expr at pc with range in [0, 1].

## How was this patch tested?

Add a new testsuite `PercentileSuite` to test percentile directly.
Updated related testcases in `ExpressionToSQLSuite`.

jiangxb1987 committed:
[SPARK-16282][SQL] Implement percentile SQL function.

SparkQA commented:
**[Test build #62088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62088/consoleFull)** for PR 14136 at commit [`1ae3df7`](https://github.com/apache/spark/commit/1ae3df7463f36ac15c4cb3138bec3bade7eae600).

SparkQA commented:
**[Test build #62088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62088/consoleFull)** for PR 14136 at commit [`1ae3df7`](https://github.com/apache/spark/commit/1ae3df7463f36ac15c4cb3138bec3bade7eae600).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Percentile(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62088/
Test FAILed.

jiangxb1987 committed:
remove unused import

SparkQA commented:
**[Test build #62089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62089/consoleFull)** for PR 14136 at commit [`d5d4fa9`](https://github.com/apache/spark/commit/d5d4fa90b77575268cf0fbbeb6c6cab058ad0bf1).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+  usage = """_FUNC_(epxr, pc) - Returns the percentile(s) of expr at pc (range: [0,1]). pc can be
+  a double or double array.""")
+case class Percentile(
+                     child: Expression,
```
Nit: Style

SparkQA commented:
**[Test build #62089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62089/consoleFull)** for PR 14136 at commit [`d5d4fa9`](https://github.com/apache/spark/commit/d5d4fa90b77575268cf0fbbeb6c6cab058ad0bf1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62089/
Test FAILed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+    aggBufferAttributes.map(_.newInstance())
+
+  override def initialize(buffer: MutableRow): Unit = {
+    for (i <- 0 until pc.size) {
```
Please use a while loop here; for is not that efficient.

jiangxb1987 replied:
Yep, I'll update that.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+
+  override def nullable: Boolean = false
+
+  override def dataType: DataType = ArrayType(DoubleType)
```
The doc says you return a Double or an Array of doubles. Which one is it?

hvanhovell replied:
nvm

jiangxb1987 replied:
Sorry...seems I left out something.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+  override val inputAggBufferAttributes: Seq[AttributeReference] =
+    aggBufferAttributes.map(_.newInstance())
+
+  override def initialize(buffer: MutableRow): Unit = {
```
Initialize counts here?

jiangxb1987 replied:
Sure, I'll move it.

hvanhovell replied:
the coins `openhashmap` will contain values of other groups if you do not initialize it here.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+    val v = child.eval(input)
+
+    v match {
+      case o: Int => counts.changeValue(o.toLong, 1L, _ + 1L)
```
We only support Ints and Longs? Why do we care about the type here? The only thing that matter is that we can sort the keys later on.

jiangxb1987 replied:
In HIVE percentile function supports only BigInt Column, so I thought we should be consistent with that.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+
+  override def dataType: DataType = ArrayType(DoubleType)
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(AnyDataType)
```
Please implement `checkInputDataTypes` in order to make sure the data type supports sorting.

jiangxb1987 replied:
I'll implement it, thanks!

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+    val maxPosition = aggreCounts.last._2 - 1
+
+    new GenericArrayData(pc.map { percentile =>
+      if (percentile < 0.0 || percentile > 1.0) {
```
We can check this at a much earlier stage.

jiangxb1987 replied:
Yep, you are right.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+    // Linear search since this won't take much time from the total execution anyway
+    // lower has the range of [0 .. total-1]
+    // The first entry with accumulated count (lower+1) corresponds to the lower position.
+    var i = 0
```
Pass an Array and use Arrays.binarySearch?

jiangxb1987 replied:
Yes, maybe binarySearch is a better way to search the position.

hvanhovell commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -536,6 +536,25 @@ object functions {
+   * @since 2.1.0
+   */
+  def percentile(e: Column, pc: Seq[Double]): Column =
+    withAggregateFunction {Percentile(e.expr, pc)}
```
Style: match the other code, e.g.: `def min(e: Column): Column = withAggregateFunction { Min(e.expr) }`

jiangxb1987 replied:
Combining that two lines will lead to a line length over 100 characters which will trigger a scalastyle check failure.

hvanhovell replied:
Would this work:
```scala
def percentile(e: Column, pc: Seq[Double]): Column = withAggregateFunction {
  Percentile(e.expr, pc)
}
```

hvanhovell commented:
@jiangxb1987 Thanks for working on this. I did a quick pass, and it is a good start. I have a few issues:

- I am a bit concerned about the memory characteristics. The worst case scenario would be that all data gets moved to a single executor. Could you document this behavior?
- Could you add support for all numeric types? 
- Could you add SQL tests as well?

hvanhovell commented:
A more performant way of this would be to plan this using a combination of count grouped by the percentile key, this percentile function. I am not sure if we should pursue that for this PR.

vectorijk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,148 @@
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ImperativeAggregate =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  var counts = new OpenHashMap[Long, Long]()
```
@jiangxb1987 I am just curious about why we use `OpenHashMap` here instead of using `mutable.Map` to correspond with code [here](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/UDAFPercentile.java#L58) in hive. Is there any specific reason?

hvanhovell replied:
OpenHashMap is typically faster and has less overhead.

vectorijk replied:
@hvanhovell Thanks!

jiangxb1987 commented:
@hvanhovell Thank you for your kindly review, the suggestions are quite useful for me. I'll try to get some time later today to update some fixes. Thanks!

jiangxb1987 committed:
1. Add support for all numeric types; 2. Add comments to document the memory characeristics; 3. Add basic SQL tests. 4. Some other refactor.

jiangxb1987 committed:
remove unused import.

SparkQA commented:
**[Test build #62313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62313/consoleFull)** for PR 14136 at commit [`4914174`](https://github.com/apache/spark/commit/491417408c51d0b5d06070d81061614e9ffcbab8).

SparkQA commented:
**[Test build #62313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62313/consoleFull)** for PR 14136 at commit [`4914174`](https://github.com/apache/spark/commit/491417408c51d0b5d06070d81061614e9ffcbab8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62313/
Test FAILed.

jiangxb1987 commented:
@hvanhovell I've fixed most of the problems mentioned above, and I also added basic tests and comments as you required. Please find some time to do a pass, thanks!

jiangxb1987 committed:
fix scala style fail.

SparkQA commented:
**[Test build #62314 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62314/consoleFull)** for PR 14136 at commit [`bf6f539`](https://github.com/apache/spark/commit/bf6f539649b33f5bb5b5e71e3b5a7dbcc77abdc3).

SparkQA commented:
**[Test build #62314 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62314/consoleFull)** for PR 14136 at commit [`bf6f539`](https://github.com/apache/spark/commit/bf6f539649b33f5bb5b5e71e3b5a7dbcc77abdc3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62314/
Test FAILed.

jiangxb1987 committed:
fix scala style fail.

SparkQA commented:
**[Test build #62315 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62315/consoleFull)** for PR 14136 at commit [`2194c9e`](https://github.com/apache/spark/commit/2194c9ea966276e242eba29eaa65cbc7b1a7bda9).

SparkQA commented:
**[Test build #62315 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62315/consoleFull)** for PR 14136 at commit [`2194c9e`](https://github.com/apache/spark/commit/2194c9ea966276e242eba29eaa65cbc7b1a7bda9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62315/
Test FAILed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,172 @@
+
+  private val percentiles: Seq[Double] = pc match {
+    case Literal(ar: GenericArrayData, _: ArrayType) =>
+      ar.asInstanceOf[GenericArrayData].array.map{ d => d.asInstanceOf[Double]}
```
Shouldn't we check here if a percentile is valid? Waiting until `eval` is really late in the game.

hvanhovell replied:
We should also check if the array is not empty.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,172 @@
+  private val percentiles: Seq[Double] = pc match {
+    case Literal(ar: GenericArrayData, _: ArrayType) =>
+      ar.asInstanceOf[GenericArrayData].array.map{ d => d.asInstanceOf[Double]}
+    case _ => Seq.empty
```
I think we should throw an error here.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,172 @@
+  override def update(buffer: MutableRow, input: InternalRow): Unit = {
+    val v = child.eval(input)
+
+    val key = v match {
```
is it faster to a pattern match for every value, or to wrap this in a function?

hvanhovell replied:
Why force everything into a double? We can also store the value. And potentially convert it to a double when interpolation is needed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,172 @@
+      case o: Float => o.toDouble
+      case o: Decimal => o.toDouble
+      case o: Double => o
+      case _ => sys.error("Percentile is restricted to Numeric types only.")
```
I don't this this is possible. Spark will try to provide numeric types (and inject casts if needed), because you have defined `override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, NumericType)`

vectorijk commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala:
```diff
@@ -475,4 +475,20 @@ class DataFrameAggregateSuite extends QueryTest with SharedSQLContext {
+      df.select(percentile($"a", 0.5d)),
+      Seq(Row(Seq.empty))
+    )
+  }
```
I think we should also give a test to see what if pass an empty array of percentiles. It may throw an exception or error.

vectorijk commented:
We also need to remove line here https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala#L240.

jiangxb1987 committed:
bugfix

jiangxb1987 committed:
add testcases.

SparkQA commented:
**[Test build #62407 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62407/consoleFull)** for PR 14136 at commit [`62324d6`](https://github.com/apache/spark/commit/62324d6712e5f2b5608dbbd3ae3f480641027005).

SparkQA commented:
**[Test build #62407 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62407/consoleFull)** for PR 14136 at commit [`62324d6`](https://github.com/apache/spark/commit/62324d6712e5f2b5608dbbd3ae3f480641027005).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62407/
Test FAILed.

jiangxb1987 committed:
fix scala style fail.

SparkQA commented:
**[Test build #62408 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62408/consoleFull)** for PR 14136 at commit [`19011ab`](https://github.com/apache/spark/commit/19011ab29983b7a7b07c6f4005c294bcdb4af1cf).

SparkQA commented:
**[Test build #62408 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62408/consoleFull)** for PR 14136 at commit [`19011ab`](https://github.com/apache/spark/commit/19011ab29983b7a7b07c6f4005c294bcdb4af1cf).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62408/
Test FAILed.

SparkQA commented:
**[Test build #62455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62455/consoleFull)** for PR 14136 at commit [`d541b46`](https://github.com/apache/spark/commit/d541b46e245dbaa200040cc9220a2095717d40ba).

SparkQA commented:
**[Test build #62455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62455/consoleFull)** for PR 14136 at commit [`d541b46`](https://github.com/apache/spark/commit/d541b46e245dbaa200040cc9220a2095717d40ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62455/
Test FAILed.

jiangxb1987 committed:
add testcases for windowing.

SparkQA commented:
**[Test build #62466 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62466/consoleFull)** for PR 14136 at commit [`6314611`](https://github.com/apache/spark/commit/63146115ad270109a75fcbfa4e608de0e7b31046).

SparkQA commented:
**[Test build #62466 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62466/consoleFull)** for PR 14136 at commit [`6314611`](https://github.com/apache/spark/commit/63146115ad270109a75fcbfa4e608de0e7b31046).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62466/
Test FAILed.

jiangxb1987 commented:
Jenkins, test this please

SparkQA commented:
**[Test build #62470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62470/consoleFull)** for PR 14136 at commit [`6314611`](https://github.com/apache/spark/commit/63146115ad270109a75fcbfa4e608de0e7b31046).

SparkQA commented:
**[Test build #62470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62470/consoleFull)** for PR 14136 at commit [`6314611`](https://github.com/apache/spark/commit/63146115ad270109a75fcbfa4e608de0e7b31046).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62470/
Test FAILed.

jiangxb1987 committed:
fix testcase fail.

SparkQA commented:
**[Test build #62516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62516/consoleFull)** for PR 14136 at commit [`510c9e0`](https://github.com/apache/spark/commit/510c9e0cba969e231f04c4100eabb69de4f40d69).

SparkQA commented:
**[Test build #62516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62516/consoleFull)** for PR 14136 at commit [`510c9e0`](https://github.com/apache/spark/commit/510c9e0cba969e231f04c4100eabb69de4f40d69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62516/
Test PASSed.

jiangxb1987 commented:
@hvanhovell Would you please review this PR? Thanks!

SparkQA commented:
**[Test build #62522 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62522/consoleFull)** for PR 14136 at commit [`510c9e0`](https://github.com/apache/spark/commit/510c9e0cba969e231f04c4100eabb69de4f40d69).

SparkQA commented:
**[Test build #62522 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62522/consoleFull)** for PR 14136 at commit [`510c9e0`](https://github.com/apache/spark/commit/510c9e0cba969e231f04c4100eabb69de4f40d69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62522/
Test PASSed.

jiangxb1987 commented:
cc @rxin @HyukjinKwon Please review this PR and tell me what should I update, thanks!

SparkQA commented:
**[Test build #63152 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63152/consoleFull)** for PR 14136 at commit [`12b9cbd`](https://github.com/apache/spark/commit/12b9cbd3002300135ac8949c7dc15cbcb7ef359e).

SparkQA commented:
**[Test build #63152 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63152/consoleFull)** for PR 14136 at commit [`12b9cbd`](https://github.com/apache/spark/commit/12b9cbd3002300135ac8949c7dc15cbcb7ef359e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63152/
Test PASSed.

SparkQA commented:
**[Test build #65036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65036/consoleFull)** for PR 14136 at commit [`6d29034`](https://github.com/apache/spark/commit/6d29034474727510225709387e30f2c1dbb4d409).

SparkQA commented:
**[Test build #65036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65036/consoleFull)** for PR 14136 at commit [`6d29034`](https://github.com/apache/spark/commit/6d29034474727510225709387e30f2c1dbb4d409).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65036/
Test PASSed.

jiangxb1987 commented:
@hvanhovell Could you look at this PR? Thank you!

hvanhovell commented:
@jiangxb1987 I think this has been superseded by https://github.com/apache/spark/commit/a18c169fd050e71fdb07b153ae0fa5c410d8de27

Could you close your PR if this is the case?

jiangxb1987 commented:
@hvanhovell In HIVE we have both of `percentile` and `percentile_approx` functions, the `percentile` function gives a precise result but requires more computing resouces, so maybe we should still support `percentile` function in Spark SQL?

rxin commented:
We definitely need this as a native implementation. One thing we should think about is memory management. collect_list, collect_set, and percentile are examples of functions that are very memory heavy and can easily OOM.

SparkQA commented:
**[Test build #66646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66646/consoleFull)** for PR 14136 at commit [`af82409`](https://github.com/apache/spark/commit/af82409f7cbebcdef126547c7cda4b8462fce631).

SparkQA commented:
**[Test build #66646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66646/consoleFull)** for PR 14136 at commit [`af82409`](https://github.com/apache/spark/commit/af82409f7cbebcdef126547c7cda4b8462fce631).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66646/
Test FAILed.

SparkQA commented:
**[Test build #3307 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3307/consoleFull)** for PR 14136 at commit [`af82409`](https://github.com/apache/spark/commit/af82409f7cbebcdef126547c7cda4b8462fce631).

SparkQA commented:
**[Test build #3307 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3307/consoleFull)** for PR 14136 at commit [`af82409`](https://github.com/apache/spark/commit/af82409f7cbebcdef126547c7cda4b8462fce631).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #67008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67008/consoleFull)** for PR 14136 at commit [`14013fd`](https://github.com/apache/spark/commit/14013fd8e90d651c6fc6db76bcd81785fab6d36f).

SparkQA commented:
**[Test build #67008 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67008/consoleFull)** for PR 14136 at commit [`14013fd`](https://github.com/apache/spark/commit/14013fd8e90d651c6fc6db76bcd81785fab6d36f).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67008/
Test FAILed.

jiangxb1987 committed:
MutableRow is replaced by InternalRow.

SparkQA commented:
**[Test build #67061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67061/consoleFull)** for PR 14136 at commit [`eb264d9`](https://github.com/apache/spark/commit/eb264d9e42b1b8e3c90ead29c74538307e369681).

SparkQA commented:
**[Test build #67061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67061/consoleFull)** for PR 14136 at commit [`eb264d9`](https://github.com/apache/spark/commit/eb264d9e42b1b8e3c90ead29c74538307e369681).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67061/
Test PASSed.

jiangxb1987 commented:
@rxin What do we need to update to make this PR accepted? Please give some advice, many thanks!

rxin commented:
I think we can merge this one first, since it is no worse than the Hive one. Then we can think about how to make it more robust.

cc @hvanhovell

rxin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala:
```diff
@@ -851,6 +851,42 @@ abstract class AggregationQuerySuite extends QueryTest with SQLTestUtils with Te
     checkAnswer(df3.groupBy().agg(covar_pop("a", "b")), Row(0.0))
   }
 
+  test("percentile") {
```
i don't think you need the test case here.

rxin replied:
basically i think this entire suite is mostly historic and pretty much obsolete now.

rxin commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala:
```diff
@@ -136,7 +136,7 @@ class HiveUDFSuite extends QueryTest with TestHiveSingleton with SQLTestUtils {
   test("SPARK-2693 udaf aggregates test") {
     checkAnswer(sql("SELECT percentile(key, 1) FROM src LIMIT 1"),
-      sql("SELECT max(key) FROM src").collect().toSeq)
+      sql("SELECT array(max(key)) FROM src").collect().toSeq)
```
what is this change about?

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -613,6 +613,46 @@ object functions {
   def min(columnName: String): Column = min(Column(columnName))
 
   /**
+   * Aggregate function: returns the exact percentile(s) of the expression in a group at pc with
```
let's not add the scala api for this for now. i'm not sure if we want to encourage users to use it -- it is super expensive.

rxin commented:
Is there a way we can add some expression level unit tests for this? I realized we don't really have existing infrastructure for unit testing aggregate expressions (only for non-aggregate expressions).

hvanhovell commented:
It is relatively straight forward to test `ImperativeAggregate`s. See:

- org.apache.spark.sql.catalyst.expressions.aggregate.HyperLogLogPlusPlusSuite
- org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentileSuite 

For some inspiration.

rxin commented:
@jiangxb1987 can you add some test cases based on Herman's comment?

jiangxb1987 commented:
@rxin @hvanhovell Thanks for your advice, I'll address them ASAP!

jiangxb1987 committed:
refactor some code; remove unnecessary test cases.

jiangxb1987 committed:
remove unnecessary test cases.

jiangxb1987 committed:
add expression level test cases.

SparkQA commented:
**[Test build #67458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67458/consoleFull)** for PR 14136 at commit [`eff3912`](https://github.com/apache/spark/commit/eff391282154d14e6b41a30d34178ffeb0e71317).

SparkQA commented:
**[Test build #67458 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67458/consoleFull)** for PR 14136 at commit [`eff3912`](https://github.com/apache/spark/commit/eff391282154d14e6b41a30d34178ffeb0e71317).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67458/
Test FAILed.

jiangxb1987 committed:
fix scala style fail.

SparkQA commented:
**[Test build #67475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67475/consoleFull)** for PR 14136 at commit [`6a41d1a`](https://github.com/apache/spark/commit/6a41d1adabd33c5b9461a2582c293994ddc36730).

SparkQA commented:
**[Test build #67475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67475/consoleFull)** for PR 14136 at commit [`6a41d1a`](https://github.com/apache/spark/commit/6a41d1adabd33c5b9461a2582c293994ddc36730).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67475/
Test PASSed.

rxin commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileSuite.scala:
```diff
@@ -0,0 +1,192 @@
+      agg.update(groupBuffer, input)
+    }
+
+    // Don't support partial aggregations for now.
```
for now? is percentile ever possible to do with partial agg?

jiangxb1987 replied:
Perhaps we could pass the HashMap `counts` as `inputBuffer` in `merge` function, then we can generate local `counts` map first, after that we merge these maps and compute the `eval()` function.

jiangxb1987 commented:
@hvanhovell I've added an test suite to test `Percentile` directly, please have a look when you have time. Thank you!

jiangxb1987 commented:
ping @hvanhovell

jiangxb1987 commented:
ping @hvanhovell

hvanhovell commented:
@jiangxb1987 like we discussed offline. Lets re-implement this as a TypedImperativeAggregate.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,201 @@
+    if (returnPercentileArray) ArrayType(DoubleType) else DoubleType
+
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(NumericType, TypeCollection(NumericType, ArrayType))
```
Just to be the devil's advocate: Why don't we support AtomicTypes here? They are all orderable.

hvanhovell replied:
The type of the second argument should be `FractionalType`. Can we also specify the type of the array (`FractionalType`)?

jiangxb1987 replied:
I'm not sure we should support AtomicTypes because not all of them could be converted to double, and we have to use double values to do interpolation operations.

hvanhovell replied:
Supporting `NumericType` does not really make sense for the percentage value. Use `FractionalType` instead.

hvanhovell replied:
BTW - you can make the analyzer add casts for you:
```scala
override def inputTypes: Seq[AbstractDataType] = percentageExpression.dataType match {
  case _: ArrayType => Seq(NumericType, ArrayType(DoubleType, false))
  case _ => Seq(NumericType, DoubleType)
}
```

Then you are alway sure you get a double or a double array for the `percentageExpression`.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,201 @@
+      return generateOutput(Seq.empty)
+    }
+
+    val sortedCounts = counts.toSeq.sortBy(_._1)(new Ordering[Number]() {
```
An atomic dataType has a ordering defined. Just use that.

jiangxb1987 committed:
rewrite Percentile using the TypedImperativeAggregate interface.

SparkQA commented:
**[Test build #69081 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69081/consoleFull)** for PR 14136 at commit [`7731066`](https://github.com/apache/spark/commit/7731066d1cc59137101f333f77194a4202522294).

SparkQA commented:
**[Test build #69081 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69081/consoleFull)** for PR 14136 at commit [`7731066`](https://github.com/apache/spark/commit/7731066d1cc59137101f333f77194a4202522294).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Countings(val counts: OpenHashMap[Number, Long]) extends Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69081/
Test FAILed.

jiangxb1987 committed:
fix class cast exception for output.

SparkQA commented:
**[Test build #69110 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69110/consoleFull)** for PR 14136 at commit [`4ace3bc`](https://github.com/apache/spark/commit/4ace3bc5402866761c6f1b61600c4d4e71321598).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69110/
Test FAILed.

jiangxb1987 commented:
retest this please.

SparkQA commented:
**[Test build #69121 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69121/consoleFull)** for PR 14136 at commit [`4ace3bc`](https://github.com/apache/spark/commit/4ace3bc5402866761c6f1b61600c4d4e71321598).

SparkQA commented:
**[Test build #69121 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69121/consoleFull)** for PR 14136 at commit [`4ace3bc`](https://github.com/apache/spark/commit/4ace3bc5402866761c6f1b61600c4d4e71321598).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69121/
Test FAILed.

jiangxb1987 committed:
Implement serializer for Percentile.

SparkQA commented:
**[Test build #69144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69144/consoleFull)** for PR 14136 at commit [`e01d0b2`](https://github.com/apache/spark/commit/e01d0b2ca5c7a0f7d693d48cb2dbefeed569fec0).

SparkQA commented:
**[Test build #69144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69144/consoleFull)** for PR 14136 at commit [`e01d0b2`](https://github.com/apache/spark/commit/e01d0b2ca5c7a0f7d693d48cb2dbefeed569fec0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class CountingsSerializer `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69144/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+    Countings()
+  }
+
+  private def evalPercentages(expr: Expression): (Boolean, Seq[Number]) = {
```
Why not return doubles?

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
+  private lazy val (returnPercentileArray: Boolean, percentages: Seq[Number]) =
```
This can be problematic with serialization. Just put the percentages in a `@transient lazy val` and inline the use of `returnPercentileArray`.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+  override def nullable: Boolean = true
+
+  override def dataType: DataType =
+    if (returnPercentileArray) ArrayType(DoubleType) else DoubleType
```
I think we should return the type of the input. We can always interpolate the value and cast that to the input type. Is this is different from what Hive does?

jiangxb1987 replied:
HIVE could return double value or array of double values even the column dataType is integer, for example:
```
hive> insert into tbl values(1,2,5,10);
hive> insert into tbl values(1),(2),(5),(10);
hive> select percentile(a, array(0, 0.25, 0.5, 0.75, 1)) from tbl;
[1.0,1.75,3.5,6.25,10.0]
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+  // Returns null for empty inputs
+  override def nullable: Boolean = true
+
+  override def dataType: DataType =
```
```scala
override lazy val dataType: DataType = percentageExpression.dataType match {
  case _: ArrayType => ArrayType(DoubleType, false)
  case _ => DoubleType
}
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+    Seq(NumericType, TypeCollection(NumericType, ArrayType))
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForNumericExpr(child.dataType, "function percentile")
```
Call `super.checkInputDataTypes()`, that will validate the inputTypes(). Also check the `percentageExpression`, that must foldable and the percentage(s) must be in the range [0, 1].

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+      counts.foreach { pair =>
+        val row = InternalRow.apply(pair._1, pair._2)
+        val unsafeRow = projection.apply(row)
+        buffer ++= unsafeRow.getBytes
```
This is extremely expensive, because you are resizing the buffer for every entry. Please use a `ByteArrayOutputStream` and a `DataOutputStream`. See this for an example: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala#L226-L239

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+
+      // Read the pairs of counts map
+      val row = new UnsafeRow(2)
+      val pairRowSizeInBytes = UnsafeRow.calculateFixedPortionByteSize(2)
```
This might cause an issue for a `DecimalType`, a decimal does not have to be fixed. I think we need to write out row sizes or not allow variable length keys. BTW if you only allow fixed length keys, you could get rid of UnsafeRows and projections and directly use a `DataOutputStream`.

hvanhovell reviewed (changes requested):
I did another pass. My main feedback is to consolidate this more in a single class.

jiangxb1987 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,292 @@
+        return Seq.empty
+      }
+
+      val sortedCounts = counts.toSeq.sortBy(_._1)(new Ordering[Number]() {
```
Maybe a dumb question: How can we order a sequence of `Number` using the `Ordering[NumericType#InternalType]` ?

hvanhovell replied:
You could cast it?

jiangxb1987 committed:
code refactor.

SparkQA commented:
**[Test build #69186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69186/consoleFull)** for PR 14136 at commit [`b0aabf9`](https://github.com/apache/spark/commit/b0aabf9824b85f1d249b25870ccda9a3a79d9691).

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,326 @@
+    Countings()
+  }
+
+  private def evalPercentages(expr: Expression): Seq[Double] = (expr.dataType, expr.eval()) match {
```
Move this to the definition of percentages. You can also make this much simpler. The analyzer guarantees that you either get a single double, or an ArrayData of double:
```scala
@transient
private lazy val percentages = percentageExpression.eval() match {
  case p: Double => Seq(p)
  case a: ArrayData => a.toDoubleArray().toSeq
}
```

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,326 @@
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
+  private lazy val returnPercentileArray = percentageExpression.dataType.isInstanceOf[ArrayType]
```
Mark it `@transient`.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,326 @@
+      defaultCheck
+    } else if (!percentageExpression.foldable) {
+      // percentageExpression must be foldable
+      TypeCheckFailure(s"The percentage(s) must be a constant literal, " +
```
Nit no string interpolation.

SparkQA commented:
**[Test build #69186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69186/consoleFull)** for PR 14136 at commit [`b0aabf9`](https://github.com/apache/spark/commit/b0aabf9824b85f1d249b25870ccda9a3a79d9691).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69186/
Test FAILed.

jiangxb1987 committed:
remove the class Countings and CountingsSerializer

SparkQA commented:
**[Test build #69188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69188/consoleFull)** for PR 14136 at commit [`5b8cd4d`](https://github.com/apache/spark/commit/5b8cd4d5ba5b2cec4e7dac45a9831303f52a84ba).

SparkQA commented:
**[Test build #69188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69188/consoleFull)** for PR 14136 at commit [`5b8cd4d`](https://github.com/apache/spark/commit/5b8cd4d5ba5b2cec4e7dac45a9831303f52a84ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

lins05 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala:
```diff
@@ -0,0 +1,262 @@
+
+    val sortedCounts = buffer.toSeq.sortBy(_._1)(
+      child.dataType.asInstanceOf[NumericType].ordering.asInstanceOf[Ordering[Number]])
+    val aggreCounts = sortedCounts.scanLeft(sortedCounts.head._1, 0L) {
```
nit: maybe `accumlatedCounts` is a slightly better name than `aggreCounts` here?

jiangxb1987 committed:
revert Percentile to accept percentages in ArrayType(NumericType).

SparkQA commented:
**[Test build #69233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69233/consoleFull)** for PR 14136 at commit [`3c699ad`](https://github.com/apache/spark/commit/3c699adfee609781c1e4ce2c08493308f5e7f511).

jiangxb1987 commented:
Currently `ImplicitTypeCasts` doesn't support cast between `ArrayType(elementType)`s, so we have to support `ArrayType(NumericType)` for now. When we have add that support, we could make  the code for analyze `percentageExpression` more concise.

SparkQA commented:
**[Test build #69233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69233/consoleFull)** for PR 14136 at commit [`3c699ad`](https://github.com/apache/spark/commit/3c699adfee609781c1e4ce2c08493308f5e7f511).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69233/
Test PASSed.

hvanhovell commented:
LGTM. Merging to master/2.1. Thanks!

rxin commented:
@hvanhovell why did this go into branch-2.1? It's way past branch cut time.

hvanhovell commented:
@rxin this is a very contained patch. It only adds the percentile function. The advantage here is that this further reduces our dependency on relatively slow Hive UDAFs (one more to go), and that we don't have to put in horrible patches like https://github.com/apache/spark/pull/16034.

HyukjinKwon commented:
Hi @rxin, I was reading related codes around this and saw - https://github.com/apache/spark/pull/14136#discussion_r84591480.

It looks many suggestions for calculating median are workarounds (e.g., https://stackoverflow.com/a/31437177).

I want to use `approximate_percentile` or `percentile` in groupby/pivot and I tried to deal with this problem for few days. I ended up with a weird code as below, e.g.,:

```python
from pyspark.sql.functions import *
from pyspark.sql.column import Column, _to_java_column


def approximate_percentile(child, percentage, accuracy=lit(10000)):
    percentile_expr = spark.sparkContext._jvm.org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
    child_expr = _to_java_column(child).expr()
    percentage_expr = _to_java_column(percentage).expr()
    accuracy_expr = _to_java_column(accuracy).expr()
    agg_func = percentile_expr(child_expr, percentage_expr, accuracy_expr)
    return Column(spark._jvm.org.apache.spark.sql.Column(agg_func.toAggregateExpression()))


df = spark.range(1)
df.select(approximate_percentile(df.id, lit(0.5))).show()
```

This code might be easily broken by Spark version as it accesses to internal packages via JVM. I use 
Scala/R but also PySpark specifically to avoid version compatibility problem (even between Spark 1.6.x and Spark 2.x) in many cases but this one now becomes flaky.

Another alternative should be to port existing logic in application side to SQL ones but I was wondering if I really should do this for single case.

It might be expensive but exposing it might also promote users to test this at least.

Could we expose this in Scala/Python/R? It should be pretty easy to expose this. Or, did I misunderstand the context and other workarounds?

cc @srowen and @zero323 who I saw answered to the questions related with this outside (e.g., stackoverflow).
